### PR TITLE
feat(nextcloud_test): Use presets for running tests

### PR DESCRIPTION
**Pull request by provokateurin**
Adds the infrastructure for https://github.com/nextcloud/neon/issues/571

### CODE REVIEW

**Comment by provokateurin** [comment]
Best to review with https://github.com/nextcloud/neon/pull/1190/files?w=1 to ignore annoying whitespace diff.

**Comment by provokateurin** [comment]
@Leptopoda Added all presets for app versions that are compatible with 27.1.

**Review by Leptopoda — CHANGES_REQUESTED** [review]
I fundamentally dislike these changes.
I can no longer run the tests from within the IDE. I agree that this is a minor disadvantage compared to what this PR brings but still something I'd like to be handled differently.

I'd rather make `nextcloud_test` an independent testing package like `flutter_test` (https://github.com/flutter/flutter/blob/master/packages/flutter_test/lib/src/test_compat.dart) by utilizing `test_api` or `test_core` if needed.

We should only need to setup a custom [`Declarer`](https://github.com/dart-lang/test/blob/master/pkgs/test_api/lib/src/backend/declarer.dart) and register all the groups to it.

File: packages/nextcloud_test/lib/src/presets.dart [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    presets[presetGroup] = versions.values.toList();
+  }
+
+  return presets;
+}();
+
+/// All tests for apps that depend on the server version must be wrapped with this method and pass along the preset.
+void presets(final String presetGroup, final dynamic Function(Preset preset) body) {
```
[the remark this answers is not in the document]
Comment (provokateurin) [reply packages/nextcloud_test/lib/src/presets.dart:50]: The tests need to be able to configure some things when creating the client like the username, so we can't do it automatically. We could do it for the container though if you like it.

File: packages/nextcloud_test/lib/src/presets.dart [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+      .whereType<Directory>()
+      .map((final d) => PathUri.parse(d.path).name);
+
+  for (final presetGroup in presetGroups) {
+    final versions = <Version?, Version>{};
+
+    final presetVersions = Directory('../nextcloud_test/docker/presets/$presetGroup')
+        .listSync(followLinks: false)
+        .whereType<File>()
+        .map((final f) => Version.parse(PathUri.parse(f.path).name));
```
[the remark this answers is not in the document]
Comment (provokateurin) [reply packages/nextcloud_test/lib/src/presets.dart:25]: I agree that it is not pretty, but using glob syntax feels very wrong in this case.

File: packages/nextcloud_test/lib/src/presets.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (Leptopoda) [reply packages/nextcloud_test/lib/src/presets.dart:50]: Ah I didn't think about this.
I don't care and we can leave it as is if you want.

**Comment by Leptopoda** [comment]
> They behave exactly the same as before.

Nope. Using either of the red marked shortcuts to run a single test result in an error similar to:
```sh
No tests match regular expression "^dashboard Get widget items v1( \(variant: .*\))?$".
```

Screenshot_20231128_123517

Running an entire file (by hitting the `run` for the main) results in:
```sh
Exception: Missing docker image ghcr.io/nextcloud/neon/dev:server-27.1.3. Please build it using ./tool/build-dev-container.sh
package:nextcloud_test/src/docker_container.dart 33:7  DockerContainer.create
===== asynchronous gap ===========================
test/core_test.dart 14:23                              main.<fn>.<fn>.<fn>
LateInitializationError: Local 'container' has not been initialized.
dart:_internal             LateError._throwLocalNotInitialized
test/core_test.dart 17:24  main.<fn>.<fn>.<fn>
```

File: packages/nextcloud_test/lib/src/presets.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (Leptopoda) [reply packages/nextcloud_test/lib/src/presets.dart:25]: :woman_shrugging:

**Comment by provokateurin** [comment]
> Nope. Using either of the red marked shortcuts to run a single test result in an error similar to:

Ok I never use it, maybe there is a way to fix it.

> Running an entire file (by hitting the run for the main) results in:

This is correct, the names of the images have changed so you need to build them first.

**Comment by provokateurin** [comment]
In IntelliJ I can still run the tests just fine, so it might be a problem with the VScode plugin?

**Comment by provokateurin** [comment]
Since the IDE debugging seems to be broken only in VScode and is probably not our fault I would like to ignore it for now (and open an issue about it to fix it later).

**Review by Leptopoda — APPROVED** [review]
Let's move on.
I still think that the ide issue must be fixed in the near future.

**Pull request merged**